text/0000-url-primitives.md

```diff
@@ -0,0 +1,353 @@
+
+export default class Router extends EmberRouter {
+  location = config.locationType;
+  rootURL = config.rootURL;
+});
+
+Router.map(...);
+```
+
+is sufficient for maintaining backwards compatibility.
+
+## How we teach this
+
+> What names and terminology work best for these concepts and why? How is this
+idea best presented? As a continuation of existing Ember patterns, or as a
+wholly new one?
+
+Like the component-manager, and modifier-manager, 
+this should be considered a low-level API that most people shouldn't need to interact with.
+Addon authors may implement different URL-handling techniques and export an URL manager 
+for app-devs to assign in the router.js file.
+
+Maybe after a bit of exploration (maybe of query params, specifically), a particular approach may be pulled in to Ember.
+
+> Would the acceptance of this proposal mean the Ember guides must be
+re-organized or altered? Does it change how Ember is taught to new users
+at any level?
+
+The guides don't need any changes, but the API documentation would need to be thorough.
+
+## Drawbacks
+
+The biggest drawback is that it would be easy to break routing in the app.
+During implementation, this could be mitigated by providing a generated route unit test that 
+symmetrically checks that toURL and fromURL are inverses of each other for the given route's expected URL / routeInfo.
+
+```ts
+import { module, test } from 'qunit';
+import { setupTest } from 'ember-qunit';
+
+module('Unit | Route | blogs/blog/posts/post', function(hooks) {
+  setupTest(hooks);
+
+  test('it exists', function(assert) {
+    let route = this.owner.lookup('route:blogs/blog/posts/post');
+    assert.ok(route);
+  });
+
+  test('urls are resolved', function(assert) {
+    let router = ;
+    let sampleUrl = '/blogs/1/posts/2';
+    let resultUrl = toURL(fromURL(sampleURL))
+
+    assert.equal(resultUrl, sampleUrl);
+
+    let mapInfo = router.mapInfoFor('blogs.blog.posts.post');
+
+    let sampleRouteInfo = {
+      mapInfo,
+      queryParams: {}
+      dynamicSegments: { blog: '1', post: '2' },
+    }
+    let resultRouteInfo = fromURL(toURL(sampleRouteInfo));
+
+    assert.deepEqual(resultRouteInfo, sampleRouteInfo);
+  });
+});
+```
+
+## Alternatives
+
+- Builtin regex matchers for the routes
+  - hard to debug
+  - can often match on incorrect portions of the Route if not thoroughly tested
+
+- Only adding QueryParams modifications (per RFC #380)
+  - not flexible enough
+  - forces a single query params implementation
+
+Prior Art: 
+- the manager patterns from elsewhere in Ember.
+
+## Unresolved questions
+
+- Should the URL Manager exist as a static config or an instantiable object per-route? The above proposal is a static config / singleton, but allowing an instance per route would allow for more varied state buckets, but could also make debugging harder as there would be an URL Manager for each route segment.
+
+- `toURL` / `fromURL` or `serialize` / `deserialize`?
```

Comment:
> `toURL` / `fromURL` or `serialize` / `deserialize`?

definitely `toURL` / `fromURL` - with `(de)serialize` you are constantly question which direction is serialize? 😇